api/envoy/api/v2/cds.proto

```diff
@@ -236,6 +236,65 @@ message Cluster {
   //   verification.
   auth.UpstreamTlsContext tls_context = 11;
 
+  // TransportSocketMatch specifies what transport socket config will be used
+  // when the matches conditions are satisfied.
+  message TransportSocketMatch {
+    // The name of the match.
+    string name = 1;
+
+    // Optional endpoint metadata match criteria.
+    // The connection to the endpoint with metadata matching what is set in this field
+    // will use the transport socket configuration specified here.
+    // The endpoints metadata entry in *envoy.transport_socket* is used to match
+    // against the values specified in this field.
+    google.protobuf.Struct match = 2;
+
+    // The configuration of the transport socket.
+    core.TransportSocket transport_socket = 3;
+  }
+
+  // Configuration to use different transport sockets for different endpoints.
+  //
+  // The entry of *envoy.transport_socket* in endpoint :ref:`Metadata <envoy_api_msg_core.Metadata>`
+  // is used to match against the transport sockets as they appear in the list. The first
+  // :ref:`match <envoy_api_msg_Cluster.TransportSocketMatch>` is used.
+  // For example, with the following match
+  //
+  // .. code-block:: json
+  //
+  // {
+  //   "transport_socket_matches": [
+  //     {
+  //       "name": "enable-mtls",
+  //       "match": { "app": "sidecar" },
```

Comment:
This match example doesn't really show why the label/metadata approach has any benefit. If there is some major win from this (e.g. involving some reduction in combinatorial blow out, etc), can we have an example here that alludes to this?

Reply:
Maybe the win is around having the same cluster on different Envoys with different transport socket matches, how would this be used in practice? I think it's worth spelling this out here, as IMHO the design is complicated and folks need to know how to use it.

Reply:
@htuch I think what you're asking is expand https://github.com/envoyproxy/envoy/pull/8100/files#diff-1d688ec4c3acf482782efa562b0142c0R293 to actual examples?

If so, I think we need two CDS response examples that explains

- EDS tells "what an endpoint is capable of": i.e. the endpoint has sidecar (or istio) means the endpoint is capable for mTLS and plaintext
- CDS specify how that is being used: CDS for Envoy A can use mTLS for endpoints with sidecar, while CDS for Envoy B keep using plaintext even the endpoints are capable with mTLS.

Reply:
Yes. I stated the reason of having match based sematics at the line  @lizan linked.

The example here originally was trying to illustrate the generic usage of the API. Totally fine to me if we need more concrete example to expand text.  just need to double check. :)

Reply:
Yes, I think just be more verbose. It doesn't even have to be a fully expansion of the CDS response, just paraphrase what @lizan has in those two bullet points.

Reply:
Done. Expand to be more verbose.